### PR TITLE
ENH enable partial_fit for forest classifiers

### PR DESCRIPTION
**Pull request by PSSF23**
#### Reference Issues/PRs

https://github.com/neurodata/scikit-tree/issues/107

### CODE REVIEW

**Review by PSSF23 — COMMENTED** [review]
`black==23.7.0` is formatting the file differently from the linting test.

**Comment by PSSF23** [comment]
@adam2392 About the `set_split_node` change you mentioned, is it still TODO?

File: sklearn/tree/_classes.py [hunk 1]
```diff
@@ -1337,8 +1335,10 @@ def partial_fit(self, X, y, classes=None, sample_weight=None, check_input=True):
                     )
 
         if X.shape[1] != self.n_features_in_:
-            msg = "Number of features %d does not match previous data %d."
-            raise ValueError(msg % (X.shape[1], self.n_features_in_))
+            raise ValueError(
+                f"X has {X.shape[1]} features, but {self.__class__.__name__} "
+                f"is expecting {self.n_features_in_} features as input."
+            )
```
Comment (adam2392) [inline sklearn/tree/_classes.py:1341]: Why do we need this change?

File: sklearn/utils/tests/test_utils.py [hunk 2]
```diff
@@ -525,13 +525,13 @@ def test_shuffle_dont_convert_to_array():
     a_s, b_s, c_s, d_s, e_s = shuffle(a, b, c, d, e, random_state=0)
 
     assert a_s == ["c", "b", "a"]
-    assert type(a_s) == list
+    assert isinstance(a_s, list)
```
Comment (adam2392) [inline sklearn/utils/tests/test_utils.py:528]: Same here? Why do we need this?

**Review by adam2392 — COMMENTED** [review]
Just a few questions to try to limit the diff wrt main

File: sklearn/tree/_classes.py (hunk 1 above)
Comment (PSSF23) [reply sklearn/tree/_classes.py:1341]: Different error message in the test.

File: sklearn/utils/tests/test_utils.py (hunk 2 above)
Comment (PSSF23) [reply sklearn/utils/tests/test_utils.py:528]: To pass linting

File: sklearn/feature_selection/tests/test_from_model.py [hunk 3]
```diff
@@ -401,10 +401,6 @@ def test_partial_fit():
     transformer.fit(np.vstack((data, data)), np.concatenate((y, y)))
     assert_array_almost_equal(X_transform, transformer.transform(data))
 
-    # check that if est doesn't have partial_fit, neither does SelectFromModel
-    transformer = SelectFromModel(estimator=RandomForestClassifier())
-    assert not hasattr(transformer, "partial_fit")
-
```
[the remark this answers is not in the document]
Comment (PSSF23) [reply sklearn/feature_selection/tests/test_from_model.py:407]: good idea

File: sklearn/utils/tests/test_utils.py (hunk 2 above)
Comment (adam2392) [reply sklearn/utils/tests/test_utils.py]: It seems like it was resolved upstream, so I just merged in changes from sklearn:main and reverted these lines for you.

Comment (adam2392) [reply sklearn/utils/tests/test_utils.py]: The motivation is to keep diff wrt sklearn:main as small as possible cuz it just makes our life easier, so when in doubt, we check upstream to make sure any fixes are already reflected there.

File: sklearn/feature_selection/tests/test_from_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adam2392) [reply sklearn/feature_selection/tests/test_from_model.py:407]: https://github.com/neurodata/scikit-learn/pull/55/commits/6c389e3f51cfe10c802b4e8a7b0fe70b1b467e28

**Comment by PSSF23** [comment]
> > `black==23.7.0` is formatting the file differently from the linting test.
> 
> Are you talking about this: [#55 (comment)](https://github.com/neurodata/scikit-learn/pull/55#issuecomment-1686731480)?
> 
> It shows an issue identified via `ruff` on my end

I have modified the format to pass the linting tests, but if I do `black sklearn/ensemble/_forest.py` again the file will still be reformatted by `black`.

**Comment by adam2392** [comment]
> > > `black==23.7.0` is formatting the file differently from the linting test.
> > 
> > 
> > Are you talking about this: [#55 (comment)](https://github.com/neurodata/scikit-learn/pull/55#issuecomment-1686731480)?
> > It shows an issue identified via `ruff` on my end
> 
> I have modified the format to pass the linting tests, but if I do `black sklearn/ensemble/_forest.py` again the file will still be reformatted by `black`.

They are frozen on black == 23.3.0 rn it seems. I would just `pip install pre-commit` to use those tools perhaps?

File: sklearn/feature_selection/tests/test_from_model.py [hunk 4]
```diff
@@ -402,7 +402,7 @@ def test_partial_fit():
     assert_array_almost_equal(X_transform, transformer.transform(data))
 
     # check that if est doesn't have partial_fit, neither does SelectFromModel
-    transformer = SelectFromModel(estimator=RandomForestClassifier())
+    transformer = SelectFromModel(estimator=RandomForestRegressor())
```
Comment (adam2392) [inline sklearn/feature_selection/tests/test_from_model.py:405]: There's an error in CI related to this: 

```
sklearn/feature_selection/tests/test_from_model.py:405:45: F821 Undefined name `RandomForestRegressor`
```

Probably need to import the RFRegressor.

**Review by adam2392 — APPROVED** [review]
Once CIs pass, it LGTM to merge.

**Comment by adam2392** [comment]
> > @adam2392 About the `set_split_node` change you mentioned, is it still TODO?
> 
> No. I fixed it in the fork and in scikit-tree. Feel free to take a look and see if it works for the forest. If so, then all the classification forests have now streaming ability.

Idk if high-priority, but it would be really cool to see how streaming works wrt ObliqueRF and MORF. Lmk if you want to discuss.

**Pull request merged**